feat: add `preserve-caught-error` rule

Amnish04 opened the pull request:
<!--
    Thank you for contributing!

    ESLint adheres to the [OpenJS Foundation Code of Conduct](https://eslint.org/conduct).
-->

#### Prerequisites checklist

- [X] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request?

<!--
    The following template is intentionally not a markdown checkbox list for the reasons
    explained in https://github.com/eslint/eslint/pull/12848#issuecomment-580302888
-->

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[X] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

<!--
    If the item you've checked above has a template, please paste the template questions below and answer them. (If this pull request is addressing an issue, you can just paste a link to the issue here instead.)
-->
fixes https://github.com/eslint/eslint/issues/19844

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/latest/contribute/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

#### What changes did you make?

I've done some work to implement an initial version of `preserve-caught-error` rule.

On top of what my plugin [already supports](https://github.com/Amnish04/eslint-plugin-error-cause/blob/main/tests/no-swallowed-error-cause.test.ts), I've made some improvements as discussed in #19844.

The rule now also:
1. Detects and flags **multiple** throw statements without a `cause` within a `CatchClause`.
2. Reports with a [special message](https://github.com/Amnish04/eslint/blob/8c3120c0fb2743db5164ceb127c81ffbd8a9bc95/lib/rules/preserve-caught-error.js#L53) when the caught error is being ignored at the parameter level. See https://github.com/Amnish04/eslint/blob/8c3120c0fb2743db5164ceb127c81ffbd8a9bc95/tests/lib/rules/preserve-caught-error.js#L188.
3. Checks for all the in-built `Error` types that support the `cause` option by default. Accepts a [customErrorTypes option](https://github.com/Amnish04/eslint/blob/8c3120c0fb2743db5164ceb127c81ffbd8a9bc95/lib/rules/preserve-caught-error.js#L35), that is a list of user-defined error types to extend what Error types to look for.
4. Offers suggestions to auto-fix instead of direct fixers.
5. All the new features are accompanies with corresponding tests.
6. Documents the rule details in `docs/src/rules/preserve-caught-error.md`.

#### Is there anything you'd like reviewers to focus on?

Happy to iterate and make any changes/fixes required :)

<!-- markdownlint-disable-file MD004 -->

Amnish04 commented:
Not sure why, I am getting CI issues with my `further_reading` links:

```
Problem writing Eleventy templates:
[11ty] 1. Having trouble writing to "./_site/rules/preserve-caught-error.html" from "./src/rules/preserve-caught-error.md" (via EleventyTemplateError)
[11ty] 2. (./src/_includes/layouts/doc.html)
[11ty]  (/home/runner/work/eslint/eslint/docs/src/_includes/components/docs-index.html)
[11ty]  (/home/runner/work/eslint/eslint/docs/src/_includes/components/search.html)
[11ty]   EleventyNunjucksError: Error with Nunjucks shortcode `link` (via Template render error)
[11ty] 3. Data missing for https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/cause
```

https://github.com/Amnish04/eslint/blob/99b3c2926152b972782a4eec8f06dea7177ccb04/docs/src/rules/class-methods-use-this.md#L4-L7

fasttime reviewed (commented):
Thanks for the pull request. Can you also add a type definition for the new rule to [lib/types/rules.d.ts](https://github.com/eslint/eslint/blob/main/lib/types/rules.d.ts)? This is a good example:

https://github.com/eslint/eslint/blob/bb370b8e79f65ee32d9d89ecf249fb74a141ad22/lib/types/rules.d.ts#L2447-L2453

After that, you can run `node tools/update-rule-type-headers.js` locally to autogenerate the TSDoc comment from the rule metadata.

fasttime commented on lib/rules/preserve-caught-error.js:
```diff
@@ -0,0 +1,263 @@
+				...customErrorTypes,
+			]);
+			return (
+				throwStatement.argument.type === "NewExpression" &&
```
Besides `NewExpression` it would be good to consider also `CallExpression` nodes in this check, in order to report on this case:

```js
try { doSomething(); } catch (err) { throw Error("Something failed"); }
```

All native JavaScript errors can be constructed without the `new` syntax.

fasttime commented on docs/src/rules/preserve-caught-error.md:
```diff
@@ -0,0 +1,87 @@
+title: preserve-caught-error
+rule_type: suggestion
+further_reading:
+- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/cause
+- https://nodejs.org/api/errors.html#errorcause
+- https://github.com/tc39/proposal-error-cause/blob/main/README.md
+- https://dev.to/amnish04/never-lose-valuable-error-context-in-javascript-3aco
```
I think it's fine to do that in this PR since this is a very small adjustment. Maybe it would be sufficient to print just the file path rather than the full URL in the GitHub repo.

fasttime commented on packages/js/src/configs/eslint-recommended.js:
```diff
@@ -76,6 +76,7 @@ module.exports = Object.freeze({
 		"no-useless-catch": "error",
 		"no-useless-escape": "error",
 		"no-with": "error",
+		"preserve-caught-error": "error",
```
We don't add new rules to the recommended config during a minor release as that is considered a breaking change. If the team agrees, we could consider adding it during the next major release. I'd suggest opening a new issue to discuss that.

What we could do in this PR is adding the new rule to [packages/eslint-config-eslint/base.js](https://github.com/eslint/eslint/blob/main/packages/eslint-config-eslint/base.js). This will enable the rule inside ESLint repos.

Amnish04 replied:
Gotcha! I've removed it from the `recommended` config, but enabled it in `eslint-plugin-eslint` config.

Amnish04 replied:
Amazing, the rule has already started doing its job!

I've fixed these issues and the CI is passing again.

For the first error, I've added an `eslint-disable-next-line` comment. The reason it was being flagged was because there was a throw statement in the catch block, but the caught error wasn't being read.

Do you think we should make sure if a `ThrowStatement` is using the return value of a function, we should check if the function is being passed the caught error?

JoshuaKGoldberg reviewed (changes requested):
A lovely start! fasttime left some very good review comments around ESLint's infrastructure that I don't have a lot to add to. I also have some separate suggestions around the options and default behavior. Cheers!

Amnish04 commented on lib/rules/preserve-caught-error.js:
```diff
@@ -0,0 +1,263 @@
+
+	create(context) {
+		const sourceCode = context.sourceCode;
+		const options = context.options?.[0] || {};
```
Ahh right, just removed the unnecessary `?`.

Amnish04 commented on tests/lib/rules/preserve-caught-error.js:
```diff
@@ -0,0 +1,215 @@
+		},
+		/* 2. Throws a new Error with unrelated cause */
+		{
+			code: `try { doSomething(); } catch (err) { const unrelated = new Error("other"); throw new Error("Something failed", { cause: unrelated }); }`,
```
Yeah I hated that as well. The only reason I did it that way was to avoid indentation discrepancies between `code` and `output`, but LLM fixed it for me!

Amnish04 commented on lib/rules/preserve-caught-error.js:
```diff
@@ -0,0 +1,263 @@
+				// Find all the Throw Statements in the catch block
+				const throwStatements = findThrowStatements(node.body);
+
+				// Check if there and throw statements and caught error is being ignored (e.g., catch { ... })
```
Fixed!

Amnish04 commented on lib/rules/preserve-caught-error.js:
```diff
@@ -0,0 +1,263 @@
+			return (
+				throwStatement.argument.type === "NewExpression" &&
+				throwStatement.argument.callee.type === "Identifier" &&
+				allErrorTypes.has(throwStatement.argument.callee.name)
```
I also like this idea as this allows us to cover all types of cases in a generic way.

But, a couple of thoughts:
1. Going with the above approach would mean cases like the following would be considered valid even though we're losing error context there. In other words, this would solve one problem (helper functions) but break the rule for multiple other important cases.
```ts
try {
	doSomething();
} catch (err) {
	if (err.code === "A") {
		throw new Error("Type A", { cause: err });
	}
	throw new TypeError("Fallback error");
}
```
2. I recently switched my approach to start with `ThrowStatement` listeners and track parent `CatchClause`, as it allows reaching every throw statement without dealing with searching tonnes of possible JavaScript constructs as discussed in https://github.com/eslint/eslint/pull/19913#discussion_r2190816862. So in order to answer the question - "If the caught error has been used at least once", we'll still have to do the manual search which we were trying to avoid. (Maybe there is a better way I can't think of)

fasttime replied:
> Proposal: I think it'd be best to omit the `customErrorTypes` logic for now. Instead, I think the rule should err on the side of permissiveness. If the `catch` block passes its caught error as an argument to _any_ call or new expression, the rule should consider the caught error as used.

By this logic, we wouldn't be able to check the error type to make sure it allows the `cause` option to be specified. Many custom errors, like Node.js' `AssertionError`, don't provide a way to specify a `cause`, and there is no point to report a violation if such an error is created without including the caught error.

I think it's better to restrict the rule on reporting specific error types only.

> ```ts
> try {
> 	doSomething();
> } catch (err) {
> 	if (err.code === "A") {
> 		throw new Error("Type A", { cause: err });
> 	}
> 	throw new TypeError("Fallback error");
> }
> ```

I think the above code should be invalid because the thrown `TypeError` has no `cause` option.

nzakas replied:
For the sake of simplicity and getting a first version of this rule, let's keep it simple and just error for the built-in error types. I agree with @fasttime that `throw new TypeError("Fallback error")` should be flagged as invalid.

Amnish04 commented on lib/rules/preserve-caught-error.js:
```diff
@@ -0,0 +1,263 @@
+				properties: {
+					customErrorTypes: {
+						type: "array",
+						items: { type: "string" },
```
Yeah, this looks like a problem. @nzakas @fasttime If you also agree, I'll remove this option for now.

fasttime replied:
I don't feel strongly about this. @nzakas what do you think?

Amnish04 replied:
Sounds good, I've removed the option for now.

We can revisit when we have something like [TypeOrValueSpecifier ](https://typescript-eslint.io/packages/type-utils/type-or-value-specifier/) implemented in core. @JoshuaKGoldberg Do we have an issue filed for it?

JoshuaKGoldberg replied:
Not that I know of. I'm also not confident it would make sense in core, since core doesn't have any concept of types.

Amnish04 commented:
@fasttime @JoshuaKGoldberg Thanks for the detailed reviews!

I have made the following changes:
1. Fixed the further reading links error message.
2. Removed the rule from recommended config (will file a separate issue once this gets landed), and enabled in `eslint-config-eslint`.
3. Fixed some edge cases that @fasttime found. See https://github.com/eslint/eslint/pull/19913#discussion_r2190832508 and https://github.com/eslint/eslint/pull/19913#discussion_r2190816862.
4. Pretty formatted the `code` and `output` in test cases, instead of having the entire code on one line.
5. Fixed some typos.

@JoshuaKGoldberg I've left some ideas on your comment. Please let me know what you think!

Amnish04 commented:
Ok, I've made some changes to handle another case for throws that use a function's return value.

It makes sure that such function accept the caught error as one of it's arguments.

Added the following test cases.

**Valid:**
```js
`try {
	doSomething();
} catch (err) {
	// passing the caught error directly
	throw createExtraneousResultsError("test", err);
}`,

	`try {
	doSomething();
} catch (err) {
	// passing the caught error as a property in an object
	throw createExtraneousResultsError("test", { err });
}`,

	`try {
	doSomething();
} catch (err) {
	// caught error passed with explicit property name
	throw createExtraneousResultsError("test", { error: err });
}`,
```

**Invalid:**
```js
/* 14. Throw uses a function's return value, but the function does not use the `cause` to construct an error. */
{
	code: `try {
	doSomething();
	} catch (err) {
		throw createExtraneousResultsError("Something went wrong");
	}`,
	errors: [
		{
			messageId: "missingCauseInFunction",
		},
	],
},
```

CI is green again, and I've rebased this on upstream `main`.

fasttime commented on tests/lib/rules/preserve-caught-error.js:
```diff
@@ -0,0 +1,454 @@
+			],
+		},
+		/* 14. Throw uses a function's return value, but the function does not use the `cause` to construct an error. */
+		{
+			code: `try {
+			doSomething();
+			} catch (err) {
+				throw createExtraneousResultsError("Something went wrong");
+			}`,
+			errors: [
+				{
+					messageId: "missingCauseInFunction",
+				},
+			],
+		},
```
I think this case should be considered valid because `createExtraneousResultsError` could return a different error type than those of built-in errors. When not sure, it's better not to report false positives. A TypeScript extension rule, with type information, could report this case as well.

nzakas replied:
Regardless of the error type, `err` is never referenced, so it seems like it should be invalid?

fasttime replied:
The main problem is that not all errors support a cause. For example, in this code, the suggestion of using the caught error as a `cause` doesn't apply:

```js
import { AssertionError } from "node:assert";

function createAssertionError(message) {
    return new AssertionError({ message });
}

try {
    testSomething();
} catch (error) {
    throw createAssertionError("Test failed");
}
```

[`DOMException`](https://developer.mozilla.org/en-US/docs/Web/API/DOMException) is another example of a common error that doesn't support a cause, as is [`SuppressedError`](https://bugzilla.mozilla.org/show_bug.cgi?id=1971000) (a built-in global as of ES2026).

nzakas replied:
Understood. My point is that if `error` is present, then I think it must be used to avoid a violation. This code could just as easily be rewritten as:

```js
import { AssertionError } from "node:assert";

function createAssertionError(message) {
    return new AssertionError({ message });
}

try {
    testSomething();
} catch {
    throw createAssertionError("Test failed");
}
```

In this case, it's clear the intent is to ignore the thrown error.

fasttime replied:
Ah, I see. Sorry for the misunderstanding. Here's another example that should be clearer:

```js
import { AssertionError } from "node:assert";

function createAssertionError(message) {
    return new AssertionError({ message });
}

try {
    testSomething();
} catch (error) {
    if (error.code !== "ENOENT") {
        throw createAssertionError("Test failed");
    }
}
```

In this case the error `error` is caught and used, but it seems legitimate not to reference it in the call expression in the `throw` statement.

If an error is caught but not used, like in my example in https://github.com/eslint/eslint/pull/19913#discussion_r2198331903, then the `no-unused-vars` should report it already ([Playground link](https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tdW51c2VkLXZhcnM6IFwiZXJyb3JcIiovXG5cbmltcG9ydCB7IEFzc2VydGlvbkVycm9yIH0gZnJvbSBcIm5vZGU6YXNzZXJ0XCI7XG5cbmZ1bmN0aW9uIGNyZWF0ZUFzc2VydGlvbkVycm9yKG1lc3NhZ2UpIHtcbiAgICByZXR1cm4gbmV3IEFzc2VydGlvbkVycm9yKHsgbWVzc2FnZSB9KTtcbn1cblxudHJ5IHtcbiAgICB0ZXN0U29tZXRoaW5nKCk7XG59IGNhdGNoIChlcnJvcikge1xuICAgIHRocm93IGNyZWF0ZUFzc2VydGlvbkVycm9yKFwiVGVzdCBmYWlsZWRcIik7XG59Iiwib3B0aW9ucyI6eyJydWxlcyI6e30sImxhbmd1YWdlT3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFGZWF0dXJlcyI6e319fX19)), so I think it's okay not to handle that case additionally in this rule.

Amnish04 replied:
Right so in the current implementation:

1. If there are no throws inside a catch block, that is a valid case.
2. If there is a throw and it does not use the cause error, that's an invalid case. This could be a function or a direct instantiation of an Error. If the function creates an error that does not support `cause` property, the rule could be disabled for that line.
3. If the `cause` property is used, but it is not same as caught error, that's an invalid case.
4. If there are throw statements, and the caught error is not being accessed, that's an invalid case.

fasttime replied:
Makes sense generally, but I'm not sure if it's a good idea to report on generic function or constructor calls. That means users will have to use a disable comment each time an error that doesn't support a `cause` is thrown in a catch block. If we want to enable the rule on arbitrary error types besides built-in errors, I think it would be better to use an inclusion/exclusion list to allow exceptions for certain error types with a central setting.

fasttime commented on tests/lib/rules/preserve-caught-error.js:
[`AggregateError`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/AggregateError/AggregateError) expects options as a third parameter, so this case should be valid:

```js
try {
    throw new Error("Original error");
} catch (error) {
    throw new AggregateError([], "Lorem ipsum", { cause: error });
}
```

this should be invalid:

```js
try {
    throw new Error("Original error");
} catch (error) {
    throw new AggregateError([], { cause: error });
}
```

Amnish04 replied:
Yes, I totally forgot about this. I've made some changes to handle the args differently specifically for `AggregateError`, but if this is too common, we might have to implement a more intelligent way. Also added a test case.

Amnish04 commented on tests/lib/rules/preserve-caught-error.js:
Ahh, nice catch!

I've fixed this use case, and added the following test case:

```js
{
	code: `try {
} catch (err) {
	{
		throw new Error();
	}
}`,
	errors: [
		{
			messageId: "missingCause",
			suggestions: [
				{
					messageId: "includeCause",
					output: `try {
} catch (err) {
	{
		throw new Error("", { cause: err });
	}
}`,
				},
			],
		},
	],
}
```

fasttime replied:
Thanks, can we also add a test case for this code?

```js
try {
} catch (err) {
	throw new AggregateError();
}
```

Amnish04 commented:
@fasttime @nzakas I've made some changes addressing recent comments. Please let me know if we need to handle some cases differently.

snitin315 commented on docs/src/rules/preserve-caught-error.md:
````diff
@@ -0,0 +1,72 @@
+
+Examples of **incorrect** code for this rule:
+
+```js
````
```suggestion
```js
/* eslint presever-caught-error: "error" */
```

snitin315 commented on docs/src/rules/preserve-caught-error.md:
````diff
@@ -0,0 +1,72 @@
+
+Examples of **correct** code for this rule:
+
+```js
````
```suggestion
```js
/* eslint presever-caught-error: "error" */
```

snitin315 commented on lib/rules/preserve-caught-error.js:
```diff
@@ -0,0 +1,307 @@
+/**
```
Right now we report the following which is a false positive:

```js
try {
    // ...
  } catch (err) {
    const opts = { cause: err }
    throw new Error("msg", { ...opts }); // Error: There is no `cause` attached to the symptom error being thrown
// or throw new Error("msg", opts)
  }
```

I think in such cases we can just ignore the node instead of reporting a false positive, because it might be a lot of work to detect if the object referred has a cause property or not.

Amnish04 replied:
Yes, thanks for noticing! I did some work to detect all types of options argument including `Identifiers`, `SpreadElements`, and `ObjectExpressions`, but it's too complicated to auto-fix such cases even if detecting is manageable.

So I dropped the idea and started ignoring such cases. I am using a custom [UNKNOWN_CAUSE](https://github.com/Amnish04/eslint/blob/1549c57c05f10032ef05fd60edc423fc33c800f4/lib/rules/preserve-caught-error.js#L49) Symbol to communicate such cases across the functions.

nzakas commented on docs/src/rules/preserve-caught-error.md:
```diff
@@ -0,0 +1,62 @@
+- https://dev.to/amnish04/never-lose-valuable-error-context-in-javascript-3aco
+---
+
+JavaScript developers often re-throw errors in `catch` blocks to add context but forget to preserve the original error, resulting in lost debugging information.
+
+Using the `cause` option when throwing new errors helps retain the **original error** and maintain complete **error chains**, which improves debuggability and traceability.
```
Please show an example here. Also, let's remove the bold. That's not necessary.

Amnish04 replied:
Sure, done!

nzakas commented on docs/src/rules/preserve-caught-error.md:
```diff
@@ -0,0 +1,62 @@
+
+## Rule Details
+
+This rule enforces the use of the [`cause`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/cause) property when throwing a new error inside a `catch` block.
+
+Checks for all built-in [error types](https://github.com/microsoft/TypeScript/blob/main/src/lib/es2022.error.d.ts) that support passing a `cause`.
```
We typically avoid linking to other resources right in the middle of our docs. They should go in "further reading".

```suggestion
This rule enforces the use of the `cause` property in a built-error error constructor when throwing a new error inside a `catch` block.
```

nzakas commented on docs/src/rules/preserve-caught-error.md:
```diff
@@ -0,0 +1,62 @@
+try {
+	doSomething();
+} catch {
+	throw new TypeError("Something went wrong");
```
Upon thinking about this further, I think this case should be valid as it seems like the intent is to throw away the error. After all, if the rule is `preserve-caught-error`, the error here isn't actually caught.

Maybe we could add an option to allow catching this pattern as well if people want?

@mdjermanovic @fasttime what do you think?

mdjermanovic replied:
> I think it's a good idea to have an option to report throw statements in catch blocks that don't bind the error, and it's fine if that defaults to `false`.

Sounds good to me.

fasttime commented on lib/rules/preserve-caught-error.js:
```diff
@@ -0,0 +1,327 @@
+		 * @param {ASTNode} node The AST node to be evaluated.
+		 * @returns {ASTNode | null } The closest parent `CatchClause` node, `null` if the `node` is not in a catch block.
+		 */
+		function findParentCatch(node) {
+			let currentNode = node;
+
+			while (currentNode && currentNode.type !== "CatchClause") {
+				currentNode = currentNode.parent;
+			}
+
+			return currentNode;
+		}
```
This case is currently being reported as a problem, but I think it shouldn't because the throw statement is not in the execution path of the catch block:

```js
try {
} catch (error) {
	foo = {
		bar() {
			throw new Error();
		}
	};
}
```

These false positives could be fixed by extending the while loop to check for nodes of type `FunctionDeclaration`, `FunctionExpression`, `ArrowFunctionExpression` or `StaticBlock`, and return `null` if one of those is found.

Amnish04 commented on docs/src/rules/preserve-caught-error.md:
@fasttime Ahh, I was always confused about the role of those fences. Thanks for clarifying, I've started using these in my docs!

Amnish04 commented on docs/src/rules/preserve-caught-error.md:
````diff
@@ -0,0 +1,116 @@
+} catch { // Can't discard the error ❌
+	// Handling and re-throw logic
+}
+```
````
Done!

Amnish04 commented on docs/src/rules/preserve-caught-error.md:
````diff
@@ -0,0 +1,116 @@
+} catch(error) { // Error is being referenced ✅
+	// Handling and re-throw logic
+}
+```
````
Thanks, done!

Amnish04 commented:
> I think `disallowUncaughtErrors` is a bit difficult to understand due to the double negative. How about `requireCatchParameter` instead?

@nzakas Agreed, I've renamed the option to `requireCatchParameter` along with accompanying tests and docs.

fasttime commented on tests/lib/rules/preserve-caught-error.js:
This looks like a false positive because the global `Error` is shadowed:

```js
import { Error } from "./my-custom-error.js";

try {
    doSomething();
} catch (error) {
    throw Error("Failed to perform error prone operations");
}
```

To catch this situation, we could extend this check in `isThrowingNewError`:

https://github.com/eslint/eslint/blob/fca4f9666eab92b8418b284b51518ee53f78f7d6/lib/rules/preserve-caught-error.js#L93-L96

with the additional condition

```js
sourceCode.isGlobalReference(throwStatement.argument.callee)
```

fasttime commented on tests/lib/eslint/eslint.js:
This is (yet another) edge case:

```js
try {
    doSomething();
} catch (error) {
    if (whatever) {
        const error = anotherError;
        throw new Error("Something went wrong");
    }
}
```

In this case, adding `{ cause: error }` to the thrown error would lead to an inconsistency, because the `error` variable in the scope of the throw statement is not the actually caught error, so it's better to not provide a suggestion here.

To detect this case, where a declaration shadows the caught error, we need to use the [Scope](https://eslint.org/docs/latest/extend/scope-manager-interface) API.

Here's a quick way to check this in the `ThrowStatement` listener, which I think should work:

```js
let scope = sourceCode.getScope(throwStatement);
do {
	const variable = scope.set.get(caughtError.name);
	if (variable) {
		break;
	}
	scope = scope.upper;
} while (scope);

if (scope?.block !== parentCatch) {
	// Caught error not visible, so do not provide suggestion...
}
```

Maybe there's a more efficient option.

Amnish04 replied:
Hmm, yeah that's a weird one. I am not sure though if this should be handled by this rule because [no-shadow](https://eslint.org/docs/latest/rules/no-shadow) would already flag the redeclaration of error.

<img width="976" height="237" alt="image" src="https://github.com/user-attachments/assets/884d546b-ed0a-443b-abcf-afd44a675ede" />

fasttime replied:
I'd say these are two distinct problems - they're not even on the same statement - so it seems a good idea to report the missing `cause` in this rule as there is no misleading suggestion.

Amnish04 replied:
So you mean even if the user adds `{ cause: error }`, we should still report it with a message stating the attached cause is not the same as caught error due to shadowing?

Amnish04 replied:
Ok, I think you mean we should still report such statements but do not provide any suggestions. However, I feel like such an implicit exception in handling would confuse the users on why the throw is being reported without suggestion, and does not get fixed even on attaching the cause.

To report this in a more specific manner, I've added another error message `caughtErrorShadowed` that is used when the attached cause is being shadowed by a closer scoped redeclaration.

Here's how it works:
```js
{
	code: `try {
		doSomething();
	} catch (error) {
		if (whatever) {
			const error = anotherError;
			throw new Error("Something went wrong", { cause: error });
		}
	}`,
	errors: [
		{
			messageId: "caughtErrorShadowed",
		},
	],
},
```

In other words, if there was no `cause` attached, we would report is as usual and suggestion would attach the caught error as `cause`. After that, another report would be made using `caughtErrorShadowed` message asking the user to fix the shadowing, although this report has no accompanying suggestions/auto-fixes.

fasttime commented on tests/lib/eslint/eslint.js:
This is also an edge case, but probably here it's okay to suggest adding `{ cause: error }`, like the rule does already.

```js
try {
    // ...
} catch (error) {
    error = foo; // overwrite caught error
    throw new Error("Boom!");
}
```

Amnish04 replied:
This is also handled by one of the existing rules:
https://eslint.org/docs/latest/rules/no-ex-assign

fasttime replied:
Even if the assignment is being reported as problem by another rule, I think we should keep reporting the missing `cause` in the throw statement as well, because these are separate problems that could be fixed individually.

fasttime commented on tests/lib/rules/preserve-caught-error.js:
```diff
@@ -0,0 +1,540 @@
+				},
+			],
+		},
+		/* 3. Throws a new Error, cause property is present but value is a different identifier */
+		/*    TODO: This should actually be a valid case since e === err */
+		{
+			code: `try {
+            doSomething();
+        } catch (err) {
+            const e = err;
+            throw new Error("Failed", { cause: e });
+        }`,
+			errors: [
+				{
+					messageId: "incorrectCause",
+					suggestions: [
+						{
+							messageId: "includeCause",
+							output: `try {
+            doSomething();
+        } catch (err) {
+            const e = err;
+            throw new Error("Failed", { cause: err });
+        }`,
+						},
+					],
+				},
+			],
+		},
```
This case is fairly hard to track, so it's okay to report it. We should keep the test case to show that it's a limitation of what the rule can do.

fasttime commented on tests/lib/rules/preserve-caught-error.js:
Currently the rule ignores those cases where the catch param is destructured in the declaration, for example:

```js
try {
	doSomething();
} catch ({ message }) {
	throw new Error(message);
}

try {
	doSomethingElse();
} catch ({ ...error }) {
	throw new Error(error.message);
}
```

I think in these cases it's safe to report the throw statements as violations, because the caught error is effectively lost. We don't need to provide a suggestion.

Amnish04 replied:
This is a confusing case as well. We decided to ignore any further checks if a caught error cannot be detected by default, i.e. when `requireCatchParameter` is disabled.

This is how we check the presence of a caught error, which I think is reasonable as it's too unnecessarily complicated to detect if a destructured `...rest` is a complete error object:
```js
const caughtError =
	parentCatch.param?.type === "Identifier"
		? parentCatch.param
		: null;
```

So in my opinion, if the user has enabled `requireCatchParameter` the entire `CatchClause` will be reported, otherwise we assume no error was caught.

fasttime replied:
I was also thinking about the `requireCatchParameter` option, and I'm not sure if that option should also apply when the catch parameter is being destructured like in the example code above:

> ```js
> try {
> 	doSomething();
> } catch ({ message }) {
> 	throw new Error(message);
> }
> 
> try {
> 	doSomethingElse();
> } catch ({ ...error }) {
> 	throw new Error(error.message);
> }
> ```


In these cases, it doesn't seem obvious that the intent was to discard the caught error, so I guess it would be good to report these cases regardless of the `requireCatchParameter` option. @nzakas what do you recommend?

Amnish04 replied:
I also feel like we should be report this as caught error is being partially lost.

I've started flagging these cases with new error messages, and provide a suggestion to refer to the entire `error` object when there are re-throws.

```js
// Reported
try {
	doSomething();
} catch ({ message, ...rest }) {
	throw new Error(message);
}

// Suggestion applied
try {
	doSomething();
} catch (error) {
	throw new Error(message);
}
```

fasttime commented on tests/lib/rules/preserve-caught-error.js:
```diff
@@ -0,0 +1,601 @@
+					suggestions: [
+						{
+							messageId: "referenceCompleteCaughtError",
+							output: `try {
+				doSomething();
+			} catch(error) {
+				throw new Error(message);
+			}`,
```
This suggestion leaves `message` undefined and introduces a new undefined variable (`error`). I think no suggestion is needed here.

Amnish04 replied:
Ok, I've stopped providing a suggestion in this case.

nzakas reviewed (dismissed):
LGTM. Thanks for all of your work on this. :pray:

Leaving open for @fasttime to review before merging.

Amnish04 commented:
> When `requireCatchParameter` is set and a throw statement is missing `cause`, the rule reports a problem on the whole catch block, once for each throw statement:
> If we are okay with reporting on the whole catch block, I think we should report only one problem in this case.
> 
> Another option would be reporting on the throw statements only, like when a catch param is provided. Here for comparison:

That makes sense, thanks for noticing! I've fixed this to individually report on all affected throws instead of reporting the catch clause multiple times.

fasttime commented on lib/rules/preserve-caught-error.js:
```diff
@@ -0,0 +1,436 @@
+										// Normal Error types
+										const optionsArg = args[1];
+
+										if (!optionsArg) {
+											// Case: `throw new Error()` → insert both message and options
+											if (args.length === 0) {
+												return fixer.replaceText(
+													throwExpression,
+													`${errorPrefix}("", { cause: ${caughtError.name} })`,
+												);
+											}
+
+											// Case: message exists, add options
+											return fixer.insertTextAfter(
+												args[0],
+												`, { cause: ${caughtError.name} }`,
+											);
+										}
```
This works well, but it removes comments and formatting inside and before the argument list. We usually like to keep those elements unchanged when possible ([here](https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tYXJyYXktY29uc3RydWN0b3I6IFwiZXJyb3JcIiovXG5cbkFycmF5KC8qIFRoaXMgY29tbWVudCAqLyk7XG5cbm5ldyBBcnJheShcbiAgMCwgLy8gemVyb1xuICAxLCAvLyBvbmVcbiAgMiAvLyB0d29cbik7XG5cbm5ldyBBcnJheSAvKiBUaGlzIGNvbW1ldCBpcyBub3QgcmVtb3ZlZCAqLyAoKTtcbiIsIm9wdGlvbnMiOnsicnVsZXMiOnt9LCJsYW5ndWFnZU9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hRmVhdHVyZXMiOnt9fX19fQ==) an example with`no-array-constructor` for comparison).

I made my attempt to keep unrelated code unchanged by inserting text only, and this is what I ended up with:

```suggestion
										const astUtils = require("./utils/ast-utils");

										if (!args[0]) {
											const lastToken = sourceCode.getLastToken(throwExpression);
											const lastCalleeToken = sourceCode.getLastToken(throwExpression.callee);
											const parenToken = sourceCode.getFirstTokenBetween(
												lastCalleeToken,
												lastToken,
												astUtils.isOpeningParenToken,
											);

											if (parenToken) {
												return fixer.insertTextAfter(
													parenToken, `"", { cause: ${caughtError.name} }`,
												);
											} else {
												return fixer.insertTextAfter(
													throwExpression.callee, `("", { cause: ${caughtError.name} })`,
												);
											}
										}

										if (!args[1]) {
											return fixer.insertTextAfter(
												args[0], `, { cause: ${caughtError.name} }`,
											);
										}
```

And similary for `AggregateError` (there are duplications that could be optimized).

fasttime reviewed (commented):
The test cases are looking good. I left some a couple of comments to improve the suggestions in certain situations.

Amnish04 commented on lib/rules/preserve-caught-error.js:
```diff
@@ -0,0 +1,505 @@
+								},
+							],
+						});
+					}
```
Oh right, I missed that. Should be fixed now!

Amnish04 commented on lib/rules/preserve-caught-error.js:
```diff
@@ -0,0 +1,505 @@
+				needsComma ? "," : ""
+			} cause: ${caughtErrorName}`;
+
+			return fixer.insertTextAfter(lastProp, insertion);
```
Oh right, the comma will just be pushed forward after the fix. Thanks for noticing!

Amnish04 commented on lib/rules/preserve-caught-error.js:
```diff
@@ -0,0 +1,505 @@
+		 * This is an indicator of an error cause node, that is too complicated to be detected and fixed.
+		 * Eg, when error options is an `Identifier` or a `SpreadElement`.
+		 */
+		const UNKNOWN_CAUSE = Symbol("unknown_cause");
```
Makes sense, I've moved this to a top level helpers section.

Amnish04 commented on lib/rules/preserve-caught-error.js:
```diff
@@ -0,0 +1,505 @@
+				"TypeError",
+				"URIError",
+				"AggregateError",
+			]);
```
Done

Amnish04 commented on lib/rules/preserve-caught-error.js:
```diff
@@ -0,0 +1,505 @@
+			}
+
+			return null;
+		}
```
Moved this as well

Amnish04 commented on lib/rules/preserve-caught-error.js:
```diff
@@ -0,0 +1,504 @@
+										 */
+										if (thrownErrorCause.name === "cause") {
+											return fixer.replaceText(
+												thrownErrorCause,
```
Wow this didn't even cross my mind, js seems to have endless quirks 😝
Just made the changes you suggested and added accompanying test cases!

Amnish04 commented on lib/rules/preserve-caught-error.js:
```diff
@@ -0,0 +1,504 @@
+				prop =>
+					prop.type === "Property" &&
+					prop.key.type === "Identifier" &&
+					prop.key.name === "cause",
```
Done! Also added a test case.

fasttime reviewed (approved):
LGTM, thanks for the patient work. Happy to see this new rule coming.
